onnxruntime/core/providers/dnnl/subgraph/dnnl_func_kernel.cc

```diff
@@ -70,6 +75,9 @@ class SubgraphPrimitive : public PrimitiveBase {
         os << "Conv-" << dnnl_node.node_index << "-";
         std::shared_ptr<DnnlConv<T>> kernel;
         kernel = std::make_shared<DnnlConv<T>>(dnnl_node, params.provider, *params.attributes, os.str());
+#ifdef ENABLE_TRAINING
+        params.provider->fwd_conv_stack.emplace(kernel);
```

Comment:
@mrry The use of the stack was due to not being able to figure out another way to get the information needed in the convolution gradient operator.

In the Onnx specification, the gradient ops are independent of the corresponding forward op. DNNL on the other hand expects to have a pointer to the forward op.

For ReluGrad and MaxPoolGrad we were able to use the inputs to find the corresponding forward op. For ConvGrad we were not able to do the same.  Ideally, we would be able to use the input weight to the ConvGrad and use that to find the corresponding Conv operator since they both have the same weight as inputs.  I was unable to find a way to actually do that.

> (1.) Is the stack always empty when calling into this method?

I went back and investigated.  If an inference run is done then the ConvGrad will not be run so the stack will not be cleared. We have added code to make sure the stack is cleared every time the graph is created. 

> (1.) Could it be a local variable instead?

The forward Conv and the backward ConvGrad will likely be in separate subgraphs. So it can not be a local variable we need something that can cross between subgraphs.

> (2.) It looks like there's an assumption that ConvGrad nodes will visited in the reverse order from the corresponding Conv nodes. However, is that guaranteed to be true if (e.g.) there is branching in the graph?

Yes, there is an assumption that the ConvGrad nodes will be visited in the reverse order from the corresponding Conv nodes.  We don't actually know if this is guaranteed to be true it has been true in our testing but admittedly we have been limited to only a few models. We are open to suggestions if you have any.

Reply:
> Yes, there is an assumption that the ConvGrad nodes will be visited in the reverse order from the corresponding Conv nodes. We don't actually know if this is guaranteed to be true it has been true in our testing but admittedly we have been limited to only a few models.

It's definitely *not* guaranteed to be true. Did you test with an Inception model, or something else that has multiple parallel branches of Conv nodes? 

As a short term patch, you could add validation to ensure that the input and weight inputs are the same for both the Conv and ConvGrad tensors, and if not raise a not-implemented error. You'd still need to try something else to handle the Inception case robustly, but it would avoid producing incorrect output in the meantime.

Reply:
> As a short term patch, you could add validation to ensure that the input and weight inputs are the same for both the Conv and ConvGrad tensors, and if not raise a not-implemented error. You'd still need to try something else to handle the Inception case robustly, but it would avoid producing incorrect output in the meantime.

Thanks for the suggestion I will implement it.  I will try and have an update by end of the day Monday. 

Do you have any suggestions on a more robust way we could get a pointer to the forward operator?

Reply:
I have added a check to the ConvGrad code that checks the tensors from the forward pass against the tensors in the backward pass. Feel free to check the code out dnnl_convgrad.h lines 194-215.

Sorry I didn't get this in on Monday as expected.